cirq-core/cirq/transformers/transformer_api.py

```diff
@@ -0,0 +1,327 @@
+# Copyright 2022 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Defines the API for circuit transformers in Cirq."""
+
+import textwrap
+import functools
+from typing import (
+    Callable,
+    Tuple,
+    TypeVar,
+    Hashable,
+    List,
+    cast,
+    Type,
+    overload,
+    TYPE_CHECKING,
+)
+import dataclasses
+import enum
+
+from cirq.circuits.circuit import CIRCUIT_TYPE
+
+if TYPE_CHECKING:
+    import cirq
+
+
+class LogLevel(enum.Enum):
+    """Different logging resolution options for `TransformerStatsLogger`.
+
+    Args:
+        ALL:     All levels. Used to filter logs when printing.
+        DEBUG:   Designates fine-grained informational events that are most useful to debug /
+                 understand in-depth any unexpected behavior of the transformer.
+        INFO:    Designates informational messages that highlight the actions of a transformer.
+        WARNING: Designates unwanted or potentially harmful situations.
+        NONE:    No levels. Used to filter logs when printing.
+    """
+
+    ALL = 0
+    DEBUG = 10
+    INFO = 20
+    WARNING = 30
+    NONE = 40
+
+
+@dataclasses.dataclass
+class LoggerNode:
+    transformer_id: int
+    transformer_name: str
+    initial_circuit: 'cirq.AbstractCircuit'
+    final_circuit: 'cirq.AbstractCircuit'
+    logs: Tuple[Tuple[LogLevel, Tuple[str, ...]], ...] = ()
+    nested_loggers: Tuple[int, ...] = ()
+
+
+class TransformerStatsLogger:
+    """Abstract Base Class for transformer logging infrastructure.
+
```

Comment:
Do we want to make a note here that this also provides the default terminal/print based logger behavior ?